README.md
# Spammer

Spammer is a CLI application that sends randomly generated email to a SMTP server.

[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/endelwar/spammer/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/endelwar/spammer/?branch=master) [![Code Coverage](https://scrutinizer-ci.com/g/endelwar/spammer/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/endelwar/spammer/?branch=master) [![Build Status](https://travis-ci.org/endelwar/spammer.svg?branch=master)](https://travis-ci.org/endelwar/spammer)

## Usage

Usage:

```
 spammer [-s|--server[="..."]] [-p|--port[="..."]] [-c|--count[="..."]] [-l|--locale[="..."]]
```

Options:

```
 --server (-s)         SMTP Server ip to send email to (default: "127.0.0.1")
 --port (-p)           SMTP Server port to send email to (default: "25")
 --count (-c)          Number of email to send (default: 10)
 --locale (-l)         Locale to use (default: "en_US")
```

## License

Copyright (c) 2014 Manuel Dalla Lana (endelwar@aregar.it)

This app is licensed under the MIT license. See `LICENSE.md`.


CHANGELOG.md
CHANGELOG
=========

2017-04-20, v0.8.0
------------------

* add locale option
* check on server port number validity
* add from and to parameter
* a lot of refactoring
* configuration for automated testing on travis-ci

2014-07-17, v0.1.0
------------------

* Initial release

tests/Bootstrap.php
<?php

/*
 * This file is part of the Spammer application.
 *
 * (c) Manuel Dalla Lana <endelwar@aregar.it>
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

error_reporting(E_ALL);
if (function_exists('date_default_timezone_set') && function_exists('date_default_timezone_get')) {
    date_default_timezone_set(@date_default_timezone_get());
}

require_once __DIR__ . '/../vendor/autoload.php';


tests/Command/SpammerCommandTest.php
<?php
namespace EndelWar\Spammer\Test\Command;

use EndelWar\Spammer\Application;
use EndelWar\Spammer\Command;
use Symfony\Component\Console\Tester\CommandTester;

class SpammerCommandTest extends \PHPUnit_Framework_TestCase
{
    protected $application;
    protected $command;

    public function setUp()
    {
        parent::setUp();

        $this->application = new Application\SpammerApplication();
        $this->command = $this->application->find('spammer');
    }

    /* testing without options */
    public function testExecute()
    {
        $commandTester = new CommandTester($this->command);
        $commandTester->execute(array());

        $this->assertRegExp('/Sending 10 email/', $commandTester->getDisplay());
        $this->assertRegExp('/Sent 10 messages/', $commandTester->getDisplay());
    }

    /* testing with option: -c 1 */
    public function testSendOneEmail()
    {
        $commandTester = new CommandTester($this->command);
        $commandTester->execute(
            array('-c' => '1')
        );

        $this->assertRegExp('/Sending 1 email/', $commandTester->getDisplay());
        $this->assertRegExp('/Sent 1 messages/', $commandTester->getDisplay());
    }

    /* testing with option: -c 5 */
    public function testSendFiveEmail()
    {
        $commandTester = new CommandTester($this->command);
        $commandTester->execute(
            array('-c' => '5')
        );

        $this->assertRegExp('/Sending 5 email/', $commandTester->getDisplay());
        $this->assertRegExp('/Sent 5 messages/', $commandTester->getDisplay());
    }

    /* testing with option: -s 127.0.0.1 */
    public function testSendUsingLocalhostIP()
    {
        $commandTester = new CommandTester($this->command);
        $commandTester->execute(
            array('-s' => '127.0.0.1')
        );

        $this->assertRegExp('/Sending 10 email using server 127.0.0.1:25/', $commandTester->getDisplay());
        $this->assertRegExp('/Sent 10 messages/', $commandTester->getDisplay());
    }

    /* testing with option: -s 192.168.1.51 */
    public function testSendUsingLocalIP()
    {
        $commandTester = new CommandTester($this->command);
        $commandTester->execute(
            array('-s' => '192.168.1.51')
        );

        $this->assertRegExp('/Sending 10 email using server 192.168.1.51:25/', $commandTester->getDisplay());
        $this->assertRegExp('/Sent 10 messages/', $commandTester->getDisplay());
    }

    /* testing with option: -p 1025 (used by mailcatcher) */
    public function testSendUsingPort1025()
    {
        $commandTester = new CommandTester($this->command);
        $commandTester->execute(
            array('-p' => '1025')
        );

        $this->assertRegExp('/Sending 10 email using server 127.0.0.1:1025/', $commandTester->getDisplay());
        $this->assertRegExp('/Sent 10 messages/', $commandTester->getDisplay());
    }
}

tests/Application/SpammerApplicationTest.php
<?php

namespace EndelWar\Spammer\Test;

use EndelWar\Spammer\Application;
use PHPUnit\Framework\TestCase;
use Symfony\Component\Console\Tester\ApplicationTester;

class SpammerApplicationTest extends TestCase
{
    /** @var Application\SpammerApplication $spammer */
    private $spammer;

    protected function setUp()
    {
        $this->spammer = new Application\SpammerApplication();
        $this->spammer->setAutoExit(false); // Set autoExit to false when testing
    }

    protected function tearDown()
    {
        $this->spammer = null;
    }

    public function testHelp()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(array('--help' => true));
        $output = $spammerTester->getDisplay();
        $this->assertContains('--server', $output);
        $this->assertContains('--port', $output);
        $this->assertContains('--count', $output);
        $this->assertContains('--to', $output);
        $this->assertContains('--from', $output);
    }

    public function testExecute()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-p' => '2500',
                '-c' => '3',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains('Sending 3 email to server 127.0.0.1:2500', $output);
        $this->assertContains('Sent 3 messages', $output);
    }

    /* testing with option: -p 2500 (used by postfix smtp-sink on travis-ci) */
    public function testSendUsingPort2500()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array('-p' => '2500')
        );

        $this->assertContains('Sending 10 email to server 127.0.0.1:2500', $spammerTester->getDisplay());
        $this->assertContains('Sent 10 messages', $spammerTester->getDisplay());
    }

    /* testing with option: -s 127.0.0.1 */
    public function testSendUsingLocalIP()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-s' => '127.0.0.1',
                '-p' => '2500',
            )
        );

        $this->assertContains('Sending 10 email to server 127.0.0.1:2500', $spammerTester->getDisplay());
        $this->assertContains('Sent 10 messages', $spammerTester->getDisplay());
    }

    /* testing with option: -c 1 */
    public function testSendOneEmail()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-s' => '127.0.0.1',
                '-p' => '2500',
                '-c' => '1',
            )
        );

        $this->assertContains('Sending 1 email', $spammerTester->getDisplay());
        $this->assertContains('Sent 1 messages', $spammerTester->getDisplay());
    }

    /* testing with option: -c 5 */
    public function testSendFiveEmail()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-s' => '127.0.0.1',
                '-p' => '2500',
                '-c' => '5',
            )
        );

        $this->assertContains('Sending 5 email', $spammerTester->getDisplay());
        $this->assertContains('Sent 5 messages', $spammerTester->getDisplay());
    }

    public function testExecuteLocalePl()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-p' => '2500',
                '-c' => '1',
                '-l' => 'pl_PL',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains('Sending 1 email to server 127.0.0.1:2500 using locale pl_PL', $output);
        $this->assertContains('Sent 1 messages', $output);
    }

    public function testWrongServerOptionLiteral()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-s' => 'localhost',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains(\InvalidArgumentException::class, $output);
        $this->assertContains('server option is not a valid IP', $output);
    }

    public function testWrongServerOptionIP()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-s' => '256.289.100.587',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains(\InvalidArgumentException::class, $output);
        $this->assertContains('server option is not a valid IP', $output);
    }

    public function testWrongCountOption()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-c' => 'Lorem',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains(\InvalidArgumentException::class, $output);
        $this->assertContains('count must be equal or greater than 1 (you want to send email, right?)', $output);
    }

    public function testWrongPortOptionTooLow()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-p' => '-123',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains(\InvalidArgumentException::class, $output);
        $this->assertContains('server port must be a number between 0 and 65536', $output);
    }

    public function testWrongPortOptionTooHigh()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-p' => '66000',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains(\InvalidArgumentException::class, $output);
        $this->assertContains('server port must be a number between 0 and 65536', $output);
    }

    public function testWrongPortOptionNaN()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-p' => 'Lorem',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains(\InvalidArgumentException::class, $output);
        $this->assertContains('server port must be a number between 0 and 65536', $output);
    }

    public function testMailerError()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-p' => '2501',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains('Connection refused', $output);
        $this->assertContains('Connection could not be established with host 127.0.0.1', $output);
    }

    public function testToAddress()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-c' => '1',
                '-p' => '2500',
                '-t' => 'user@example.org',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains('Sending 1 email to server 127.0.0.1:2500 to user@example.org using locale en_US', $output);
        $this->assertContains('Sent 1 messages', $output);
    }

    public function testToFqdn()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-c' => '1',
                '-p' => '2500',
                '-t' => 'example.org',
            )
        );
        $output = $spammerTester->getDisplay();

        $this->assertContains('Sending 1 email to server 127.0.0.1:2500 to example.org using locale en_US', $output);
        $this->assertRegExp('/Sending email nr. 1: \S+@\S+\.\S{2,} => \S+@example.org/', $output);
        $this->assertContains('Sent 1 messages', $output);
    }

    public function testFromAddress()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-c' => '1',
                '-p' => '2500',
                '-f' => 'user@example.org',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains('Sending 1 email to server 127.0.0.1:2500 from user@example.org using locale en_US', $output);
        $this->assertContains('Sent 1 messages', $output);
    }

    public function testFromFqdn()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-c' => '1',
                '-p' => '2500',
                '-f' => 'example.org',
            )
        );
        $output = $spammerTester->getDisplay();

        $this->assertContains('Sending 1 email to server 127.0.0.1:2500 from example.org using locale en_US', $output);
        $this->assertRegExp('/Sending email nr. 1: \S+@example.org => \S+@\S+\.\S{2,}/', $output);
        $this->assertContains('Sent 1 messages', $output);
    }

    public function testFromAddressToAddress()
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-c' => '1',
                '-p' => '2500',
                '-f' => 'user1@example.org',
                '-t' => 'user2@example.com',
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains('Sending 1 email to server 127.0.0.1:2500 from user1@example.org to user2@example.com using locale en_US', $output);
        $this->assertContains('Sending email nr. 1: user1@example.org => user2@example.com', $output);
        $this->assertContains('Sent 1 messages', $output);
    }

    /**
     * @param $invalidFromTo
     *
     * @dataProvider badFromToProvider
     */
    public function testBadFromTo($invalidFromTo)
    {
        $spammerTester = new ApplicationTester($this->spammer);
        $spammerTester->run(
            array(
                '-c' => '1',
                '-p' => '2500',
                '-f' => $invalidFromTo,
            )
        );
        $output = $spammerTester->getDisplay();
        $this->assertContains(\InvalidArgumentException::class, $output);
        $this->assertContains('to and from must be a valid email address or a FQDN', $output);
    }

    public function badFromToProvider()
    {
        return [
            'boolean' => [true],
            'string' => ['abcd'],
            'email: single char' => ['@'],
            'email: only domain' => ['@test'],
            'email: localhost' => ['me@localhost'],
            'email: double dot' => ['test@example..com'],
            'email: dot at end' => ['test@example.com.'],
            'email: dot at start' => ['.test@example.com'],
            'email: double @' => ['test@test@example.com'],
            'domain: dot at end' => ['notvalid.com.'],
            'domain: dot at start' => ['.notvalid.com'],
            'domain: minus at start' => ['-notvalid.com'],
            'domain: minus at end' => ['notvalid.com-'],
        ];
    }
}


src/Application/SpammerApplication.php
<?php

namespace EndelWar\Spammer\Application;

use EndelWar\Spammer\Command;
use Symfony\Component\Console\Application;
use Symfony\Component\Console\Input\InputInterface;

class SpammerApplication extends Application
{
    /**
     * Gets the name of the command based on input.
     *
     * @param InputInterface $input The input interface
     *
     * @return string The command name
     */
    protected function getCommandName(InputInterface $input)
    {
        return 'spammer';
    }

    /**
     * Gets the default commands that should always be available.
     *
     * @return array An array of default Command instances
     * @throws \Symfony\Component\Console\Exception\LogicException
     */
    protected function getDefaultCommands()
    {
        // Keep the core default commands to have the HelpCommand
        // which is used when using the --help option
        $defaultCommands = parent::getDefaultCommands();

        $defaultCommands[] = new Command\SpammerCommand();

        return $defaultCommands;
    }

    /**
     * Overridden so that the application doesn't expect the command
     * name to be the first argument.
     */
    public function getDefinition()
    {
        $inputDefinition = parent::getDefinition();
        // clear out the normal first argument, which is the command name
        $inputDefinition->setArguments();

        return $inputDefinition;
    }
}


src/Command/SpammerCommand.php
<?php

namespace EndelWar\Spammer\Command;

use Faker;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Formatter\OutputFormatterStyle;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;

class SpammerCommand extends Command
{
    protected function configure()
    {
        $count = 10;
        $smtpServerIp = '127.0.0.1';
        $smtpServerPort = '25';
        $locale = 'en_US';

        $this
            ->setName('spammer')
            ->setDescription('Send random content email')
            ->setDefinition(
                array(
                    new InputOption('server', 's', InputOption::VALUE_OPTIONAL, 'SMTP Server ip to send email to', $smtpServerIp),
                    new InputOption('port', 'p', InputOption::VALUE_OPTIONAL, 'SMTP Server port to send email to', $smtpServerPort),
                    new InputOption('count', 'c', InputOption::VALUE_OPTIONAL, 'Number of email to send', $count),
                    new InputOption('locale', 'l', InputOption::VALUE_OPTIONAL, 'Locale to use', $locale),
                    new InputOption('to', 't', InputOption::VALUE_OPTIONAL, 'To address or domain'),
                    new InputOption('from', 'f', InputOption::VALUE_OPTIONAL, 'From address or domain'),
                )
            );
    }

    /**
     * @param InputInterface $input
     * @param OutputInterface $output
     * @return integer
     * @throws \InvalidArgumentException
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $validInput = $this->validateInput($input);

        $style = new OutputFormatterStyle('green', null, array('bold'));
        $output->getFormatter()->setStyle('bold', $style);
        $output->writeln('<comment>Spammer starting up</comment>');
        $message = '<info>Sending </info><bold>' . $validInput['count'] . '</bold>' .
            '<info> email to server </info><bold>' . $validInput['smtpServerIp'] . '</bold>' .
            '<info>:</info><bold>' . $validInput['smtpServerPort'] . '</bold>';
        if ($validInput['from'] !== '') {
            $message .= '<info> from </info><bold>' . $validInput['from'] . '</bold>';
        }
        if ($validInput['to'] !== '') {
            $message .= '<info> to </info><bold>' . $validInput['to'] . '</bold>';
        }
        $output->write($message);
        $output->writeln('<info> using locale </info><bold>' . $validInput['locale'] . '</bold>');

        $faker = Faker\Factory::create($validInput['locale']);
        $faker->seed(mt_rand());

        $transport = \Swift_SmtpTransport::newInstance()->setHost($validInput['smtpServerIp'])->setPort(
            $validInput['smtpServerPort']
        );
        $mailer = \Swift_Mailer::newInstance($transport);

        $numSent = 0;
        for ($i = 0; $i < $validInput['count']; $i++) {
            $emaiText = $faker->realText(mt_rand(200, 1000));
            $email_subject = implode(' ', $faker->words(mt_rand(3, 7)));
            $message = \Swift_Message::newInstance($email_subject);

            $from = $this->getFromTo($faker, $validInput['from']);
            $message->setFrom($from);

            $to = $this->getFromTo($faker, $validInput['to']);
            $message->setTo($to);

            $message->setBody($emaiText, 'text/plain');
            $message->addPart('<p>' . $emaiText . '</p>', 'text/html');

            $output->writeln('Sending email nr. ' . ($i + 1) . ': ' . key($from) . ' => ' . key($to));
            try {
                $numSent += $mailer->send($message);
            } catch (\Swift_TransportException $swe) {
                $output->writeln('<error>' . $swe->getMessage() . '</error>');

                return 1;
            }
        }

        $output->writeln('Sent ' . $numSent . ' messages');

        unset($faker);

        return 0;
    }

    /**
     * @param $faker
     * @param $validInputFromTo
     * @return array
     */
    private function getFromTo($faker, $validInputFromTo)
    {
        // generate fake address and name if null
        if ($validInputFromTo === '') {
            return [$faker->safeEmail => $faker->name];
        }

        // use user submitted email if is email
        if (filter_var($validInputFromTo, FILTER_VALIDATE_EMAIL)) {
            return [$validInputFromTo => $faker->name];
        }

        // get a random username and attach it to domain supplied by user
        $user = strstr($faker->safeEmail, '@', true);

        return [$user . '@' . $validInputFromTo => $faker->name];
    }

    /**
     * @param InputInterface $input
     * @throws \InvalidArgumentException
     * @return array
     */
    protected function validateInput(InputInterface $input)
    {
        $validInput = array();
        $validInput['smtpServerIp'] = $input->getOption('server');
        $this->validateInputServerIP($validInput['smtpServerIp']);

        $validInput['smtpServerPort'] = $input->getOption('port');
        $this->validateInputServerPort($validInput['smtpServerPort']);

        $validInput['count'] = (int)$input->getOption('count');
        $this->validateInputCount($validInput['count']);

        $validInput['locale'] = $input->getOption('locale');

        $validInput['to'] = $this->validateInputToFrom($input->getOption('to'));

        $validInput['from'] = $this->validateInputToFrom($input->getOption('from'));

        return $validInput;
    }

    /**
     * @param string $ip
     * @throws \InvalidArgumentException
     */
    private function validateInputServerIP($ip)
    {
        if (!filter_var($ip, FILTER_VALIDATE_IP)) {
            throw new \InvalidArgumentException('server option is not a valid IP');
        }
    }

    /**
     * @param int $port
     * @throws \InvalidArgumentException
     */
    private function validateInputServerPort($port)
    {
        if (!is_numeric($port) || ($port < 0 || $port > 65535)) {
            throw new \InvalidArgumentException('server port must be a number between 0 and 65536');
        }
    }

    /**
     * @param int $count
     * @throws \InvalidArgumentException
     */
    private function validateInputCount($count)
    {
        if ($count < 1) {
            throw new \InvalidArgumentException('count must be equal or greater than 1 (you want to send email, right?)');
        }
    }

    /**
     * @param $string
     * @return string
     * @throws \InvalidArgumentException
     */
    private function validateInputToFrom($string)
    {
        if (null === $string) {
            return '';
        }

        $string = strtolower($string);
        if (strpos($string, '@') !== false) {
            if (filter_var($string, FILTER_VALIDATE_EMAIL)) {
                return $string;
            }
        } else {
            if ($this->filter_var_domain($string)) {
                return $string;
            }
        }

        throw new \InvalidArgumentException('to and from must be a valid email address or a FQDN');
    }

    /**
     * @param $domain
     * @return bool|mixed
     */
    private function filter_var_domain($domain)
    {
        $domain = strtolower($domain);
        $regex = "/^((?=[a-z0-9-]{1,63}\.)(xn--)?[a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,63}$/";

        return preg_match($regex, $domain);
    }
}
